executor/analyze.go

```diff
@@ -502,16 +533,450 @@ type AnalyzeFastExec struct {
+	e.cache = store.GetRegionCache()
+	startKey, endKey := tablecodec.GetTableHandleKeyRange(e.PhysicalTableID)
+	var loc *tikv.KeyLocation
+	var err error
+	// extract all regions contain the table
+	e.scanTasks = e.scanTasks[:0]
+	e.sampTasks = e.sampTasks[:0]
+	// If length of KeyLocation.StartKey is 0, then the start key is infinitesimal.
+	// If length of KeyLocation.EndKey is 0, then the end kety is infinity.
+	for loc, err = e.cache.LocateKey(bo, startKey); bytes.Compare(loc.StartKey, endKey) <= 0 && err == nil; loc, err = e.cache.LocateKey(bo, loc.EndKey) {
+		if bytes.Compare(endKey, loc.EndKey) < 0 || bytes.Compare(loc.StartKey, startKey) < 0 || len(loc.StartKey) == 0 || len(loc.EndKey) == 0 {
+			e.scanTasks = append(e.scanTasks, loc)
+			if len(loc.StartKey) == 0 || bytes.Compare(loc.StartKey, startKey) < 0 {
+				loc.StartKey = startKey
+			}
+			if len(loc.EndKey) == 0 || bytes.Compare(endKey, loc.EndKey) < 0 {
+				loc.EndKey = endKey
+				break
+			}
+		} else {
+			e.sampLocs <- loc
+		}
+	}
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	close(e.sampLocs)
+	e.wg.Wait()
+	for i := 0; i < e.concurrency; i++ {
+		if errs[i] != nil {
+			return false, errors.Trace(errs[i])
+		}
+	}
+	for i := 0; i < e.concurrency; i++ {
+		if needRebuildForRoutine[i] == true {
+			return true, nil
+		}
+	}
+	for i := 0; i < e.concurrency; i++ {
+		e.sampTasks = append(e.sampTasks, sampTasksForRoutine[i]...)
+	}
+
+	return false, nil
+}
+
+func (e *AnalyzeFastExec) decodeValues(sValue []byte) (values map[int64]types.Datum, err error) {
+	colID2FieldTypes := make(map[int64]*types.FieldType, len(e.colsInfo))
+	if e.pkInfo != nil {
+		colID2FieldTypes[e.pkInfo.ID] = &e.pkInfo.FieldType
+	}
+	for _, col := range e.colsInfo {
+		colID2FieldTypes[col.ID] = &col.FieldType
+	}
+	values, err = tablecodec.DecodeRow(sValue, colID2FieldTypes, e.ctx.GetSessionVars().Location())
+	return values, errors.Trace(err)
+}
+
+func (e *AnalyzeFastExec) getValueByInfo(colInfo *model.ColumnInfo, values map[int64]types.Datum) (types.Datum, error) {
+	val, ok := values[colInfo.ID]
+	if !ok {
+		return table.GetColOriginDefaultValue(e.ctx, colInfo)
+	}
+	return val, nil
+}
+
+func (e *AnalyzeFastExec) updateCollectorSamples(collectors []*statistics.SampleCollector, sValue []byte, sKey kv.Key, samplePos int32, hasPKInfo int) (err error) {
+	// Decode the cols value in order.
+	var values map[int64]types.Datum
+	values, err = e.decodeValues(sValue)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	// Update the primary key collector.
+	if hasPKInfo > 0 {
+		v, ok := values[e.pkInfo.ID]
+		if !ok {
+			var key int64
+			_, key, err = tablecodec.DecodeRecordKey(sKey)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			v = types.NewIntDatum(key)
+		}
+		if collectors[0].Samples[samplePos] == nil {
+			collectors[0].Samples[samplePos] = &statistics.SampleItem{}
+		}
+		collectors[0].Samples[samplePos].Ordinal = int(samplePos)
```

Comment:
`Ordinal` represents this item's relative order of physical position among samples.
For example, if there are two regions, they have condition that `region1 < region2`.
And you sample item1 from region1 and item2, item3 from region2 and they have condition that `item1.key < item2.key < item3.key`.
Then you have to meet condition that `item1.Ordinal < item2.Ordinal < item3.Ordinal`.
But in this implementation, `item2.Ordinal < item3.Ordinal < item1.Ordinal` can happen.
@eurekaka PTAL and please help to confirm if this problem exist?